### PR TITLE
Add a .mllib -> .cmxs rule

### PR DESCRIPTION
**Pull request by Armael**
Attempt at fixing #131 . Note that I'm a complete stranger to ocamlbuild rules — does this look like what one would expect?

### CODE REVIEW

**Comment by Armael** [comment]
I added a test checking that the rule doesn't trigger if a .mldylib is present (and a Changes entry).

**Comment by gasche** [comment]
(Did you check that the test does fail if you reverse the rule declaration order?)

**Comment by Armael** [comment]
Yes.

More precisely, it fails on the following diff, or if I remove the .mldylib from the tree in the test case.
```
diff --git a/src/ocaml_specific.ml b/src/ocaml_specific.ml
index 627ba6d..bfa1b76 100644
--- a/src/ocaml_specific.ml
+++ b/src/ocaml_specific.ml
@@ -351,18 +351,6 @@ rule "ocaml: cmx & o -> cmxa & a"
         transitivitely."
   (Ocaml_compiler.native_library_link "%.cmx" "%.cmxa");;

-rule "ocaml: mldylib & p.cmx* & p.o* -> p.cmxs & p.so"
-  ~prods:["%.p.cmxs"; x_p_dll]
-  ~dep:"%.mldylib"
-  (Ocaml_compiler.native_profile_shared_library_link_mldylib "%.mldylib" "%.p.cmxs");;
-
-rule "ocaml: mldylib & cmx* & o* -> cmxs & so"
-  ~prods:["%.cmxs"; x_dll]
-  ~dep:"%.mldylib"
-  ~doc:"Builds a .cmxs (native archive for dynamic linking) containing exactly \
-        the modules listed in the corresponding .mldylib file."
-  (Ocaml_compiler.native_shared_library_link_mldylib "%.mldylib" "%.cmxs");;
-
 rule "ocaml: mllib & p.cmx* & p.o* -> p.cmxs & p.so"
   ~prods:["%.cmxs"; x_p_dll]
   ~dep:"%.mllib"
@@ -376,6 +364,18 @@ rule "ocaml: mllib & cmx* & o* -> cmxs & so"
         could be found."
   (Ocaml_compiler.native_shared_library_link_mldylib "%.mllib" "%.cmxs");;

+rule "ocaml: mldylib & p.cmx* & p.o* -> p.cmxs & p.so"
+  ~prods:["%.p.cmxs"; x_p_dll]
+  ~dep:"%.mldylib"
+  (Ocaml_compiler.native_profile_shared_library_link_mldylib "%.mldylib" "%.p.cmxs");;
+
+rule "ocaml: mldylib & cmx* & o* -> cmxs & so"
+  ~prods:["%.cmxs"; x_dll]
+  ~dep:"%.mldylib"
+  ~doc:"Builds a .cmxs (native archive for dynamic linking) containing exactly \
+        the modules listed in the corresponding .mldylib file."
+  (Ocaml_compiler.native_shared_library_link_mldylib "%.mldylib" "%.cmxs");;
+
 rule "ocaml: p.cmx & p.o -> p.cmxs & p.so"
   ~prods:["%.p.cmxs"; x_p_dll]
   ~deps:["%.p.cmx"; x_p_o]

```

File: Changes [hunk 1]
```diff
@@ -26,6 +26,9 @@ NEXT_RELEASE:
 - #127: install ocamlbuild's man pages, missing since 4.02
   (Gabriel Scherer)
 
+- #132: add a rule producing .cmxs from .mllib files
+  (Armaël Guéneau, requested by Daniel Bünzli)
```
Comment (gasche) [inline Changes]: Could you add just a bit more information here for people that have not followed the discussion? Something like "(Previously `.mllib` would only produce `.cm{x,}a`, and you had to use another file `.mldylib` for `.cmxs`. Now, a `.mllib` can also produce a `.cmxs`, but `.mldylib` files still take precedence when they exist.)"

Comment (Armael) [inline Changes]: Done (I think).

**Comment by gasche** [comment]
Now I'm thinking of something else that we could test: if the `.mllib` file has the same name as a `.cmx` file, which of the two will take precedence to build a `.cmxs`? We want the `.mllib` file to take precedence, I suppose. Could you add a test for this? (Having a test for the `.mldylib` case would also make sense.)

**Comment by dbuenzli** [comment]
That was quick, thanks Armael !

**Comment by Armael** [comment]
Added in https://github.com/ocaml/ocamlbuild/commit/69ef32cf7c77ec71bf2dbafb871d9534f72c42bb

**Comment by gasche** [comment]
(It seems that there is a complication, see the discussion in #131 again.)

**Pull request merged**